File: solver.go
package main

import "fmt"

// Play structure used to play every round
type Play struct {
	value, row, col int
}

// Sudoku game structure that take the game as an argument
type Sudoku struct {
	stack []Play
	board [9][9]int
}

// Finds th next empty cell on the Sudoku boards
func (s *Sudoku) findEmptyCell() (int, int) {
	for row := 0; row < 9; row++ {
		for col := 0; col < 9; col++ {
			if s.board[row][col] == 0 {
				return row, col // row, col
			}
		}
	}
	return -1, -1
}

// Try a valid value
func (s *Sudoku) tryValue(row, col int) bool {
	for val := s.board[row][col] + 1; val <= 9; val++ {
		status := false

		if s.checkColumn(col, val) {
			if s.checkRow(row, val) {
				if s.checkBlock(row, col, val) {
					status = true
				}
			}
		}

		if status {
			s.board[row][col] = val
			s.stack = append(s.stack, Play{val, row, col})
			return true
		}
	}
	return false
}

// Check if val
func (s *Sudoku) checkColumn(col, val int) bool {
	for row := 0; row < 9; row++ {
		if s.board[row][col] == val {
			return false
		}
	}
	return true
}

func (s *Sudoku) checkRow(row, val int) bool {
	for col := 0; col < 9; col++ {
		if s.board[row][col] == val {
			return false
		}
	}
	return true
}

func (s *Sudoku) checkCells(rowL, rowH, colL, colH, val int) bool {
	for i := rowL; i < rowH; i++ {
		for j := colL; j < colH; j++ {
			if s.board[i][j] == val {
				return false
			}
		}
	}
	return true
}

func (s *Sudoku) checkBlock(row, col, val int) bool {
	if row < 3 {
		if col < 3 {
			return s.checkCells(0, 3, 0, 3, val)
		} else if 3 <= col && col <= 5 {
			return s.checkCells(0, 3, 3, 6, val)
		} else if 5 < col {
			return s.checkCells(0, 3, 6, 9, val)
		}
	} else if 3 <= row && row <= 5 {
		if col < 3 {
			return s.checkCells(3, 6, 0, 3, val)
		} else if 3 <= col && col <= 5 {
			return s.checkCells(3, 6, 3, 6, val)
		} else if 5 < col {
			return s.checkCells(3, 6, 6, 9, val)
		}
	} else if 5 < row {
		if col < 3 {
			return s.checkCells(6, 9, 0, 3, val)
		} else if 3 <= col && col <= 5 {
			return s.checkCells(6, 9, 3, 6, val)
		} else if 5 < col {
			return s.checkCells(6, 9, 6, 9, val)
		}
	}
	return false
}

func (s *Sudoku) backtrack() {
	backtrackStatus := false
	for !backtrackStatus {
		var val Play
		val, s.stack = s.stack[len(s.stack)-1], s.stack[:len(s.stack)-1]
		backtrackStatus = s.tryValue(val.row, val.col)
		if !backtrackStatus {
			s.board[val.row][val.col] = 0
		}
	}
}

func (s *Sudoku) TestSolve() {
	row, col := 0, 0

	for row != -1 {
		// find an emty cell
		row, col = s.findEmptyCell()
		// try a value in this cell
		if row != -1 {
			if s.tryValue(row, col) {
				continue
			} else {
				s.backtrack()
			}
		} else {
			break
		}
	}

}

func (s *Sudoku) printBoard() {
	for i := 0; i < 9; i++ {
		if (i%3) == 0 && i != 0 {
			fmt.Println("- - - - - - - - - - -")
		}

		for j := 0; j < 9; j++ {
			if (j%3) == 0 && j != 0 {
				fmt.Printf("| ")
			}
			if j == 8 {
				fmt.Printf("%d\n", s.board[i][j])
			} else {
				fmt.Printf("%d ", s.board[i][j])
			}
		}
	}
	fmt.Printf("\n")
}

func main() {
	theGame := [9][9]int{{8, 0, 0, 0, 0, 0, 0, 0, 0},
		{0, 0, 3, 6, 0, 0, 0, 0, 0},
		{0, 7, 0, 0, 9, 0, 2, 0, 0},
		{0, 5, 0, 0, 0, 7, 0, 0, 0},
		{0, 0, 0, 0, 4, 5, 7, 0, 0},
		{0, 0, 0, 1, 0, 0, 0, 3, 0},
		{0, 0, 1, 0, 0, 0, 0, 6, 8},
		{0, 0, 8, 5, 0, 0, 0, 1, 0},
		{0, 9, 0, 0, 0, 0, 4, 0, 0}}
	game := Sudoku{board: theGame}
	game.printBoard()
	game.TestSolve()
	game.printBoard()
}

File: solver.py
# game = [[5, 3, 0, 0, 7, 0, 0, 0, 0],
#         [6, 0, 0, 1, 9, 5, 0, 0, 0],
#         [0, 9, 8, 0, 0, 0, 0, 6, 0],
#         [8, 0, 0, 0, 6, 0, 0, 0, 3],
#         [4, 0, 0, 8, 0, 3, 0, 0, 1],
#         [7, 0, 0, 0, 2, 0, 0, 0, 6],
#         [0, 6, 0, 0, 0, 0, 2, 8, 0],
#         [0, 0, 0, 4, 1, 9, 0, 0, 5],
#         [0, 0, 0, 0, 8, 0, 0, 7, 9]]
game = [[8, 0, 0, 0, 0, 0, 0, 0, 0],
        [0, 0, 3, 6, 0, 0, 0, 0, 0],
        [0, 7, 0, 0, 9, 0, 2, 0, 0],
        [0, 5, 0, 0, 0, 7, 0, 0, 0],
        [0, 0, 0, 0, 4, 5, 7, 0, 0],
        [0, 0, 0, 1, 0, 0, 0, 3, 0],
        [0, 0, 1, 0, 0, 0, 0, 6, 8],
        [0, 0, 8, 5, 0, 0, 0, 1, 0],
        [0, 9, 0, 0, 0, 0, 4, 0, 0]]


class Play:
    def __init__(self, value, row, col):
        self.value = value
        self.row = row
        self.col = col


class Sudoku:
    def __init__(self, board):
        self.stack = []
        self.board = board

    def find_empty_cell(self):
        for i in range(9):
            for j in range(9):
                if self.board[i][j] == 0:
                    return i, j  # row, col
        return None, None

    """OPTIMISE"""
    def try_value(self, row, col):
        for val in range(self.board[row][col] + 1, 10):
            status = all([self.check_column(col, val),
                          self.check_row(row, val),
                          self.check_block(row, col, val)])
            if status:
                self.board[row][col] = val
                self.stack.append(Play(val, row, col))
                return True
        return False

    # speed up by exiting on false
    def check_column(self, col, val):
        return all(row[col] != val for row in self.board)

    # speed up by exiting on false
    def check_row(self, row, val):
        return all(col != val for col in self.board[row])

    def check_cells(self, row_l, row_h, col_l, col_h, val):
        for i in range(row_l, row_h):
            for j in range(col_l, col_h):
                if self.board[i][j] is val:
                    return False
        return True

    """OPTIMISE"""
    def check_block(self, row, col, val):
        if row < 3 and col < 3:  # 1
            return self.check_cells(0, 3, 0, 3, val)
        elif row < 3 <= col <= 5:  # 2
            return self.check_cells(0, 3, 3, 6, val)
        elif row < 3 and col > 5:  # 3
            return self.check_cells(0, 3, 6, 9, val)
        elif col < 3 <= row <= 5:  # 4
            return self.check_cells(3, 6, 0, 3, val)
        elif 3 <= row <= 5 and 3 <= col <= 5:  # 5
            return self.check_cells(3, 6, 3, 6, val)
        elif 3 <= row <= 5 < col:  # 6
            return self.check_cells(3, 6, 6, 9, val)
        elif row > 5 and col < 3:  # 7
            return self.check_cells(6, 9, 0, 3, val)
        elif 3 <= col <= 5 < row:  # 8
            return self.check_cells(6, 9, 3, 6, val)
        elif row > 5 and col > 5:  # 9
            return self.check_cells(6, 9, 6, 9, val)

    def backtrack(self):
        backtrack_status = False
        while not backtrack_status:
            val = self.stack.pop(-1)
            backtrack_status = self.try_value(val.row, val.col)
            if backtrack_status is False:
                self.board[val.row][val.col] = 0

    def solve(self):
        row, col = 0, 0

        while row is not None and col is not None:
            # find an empty cell
            row, col = self.find_empty_cell()
            # try a value in cell
            if row is not None and col is not None:
                if self.try_value(row, col):
                    continue
                else:
                    self.backtrack()
            else:
                break

    def print_board(self):
        for i in range(9):
            if i % 3 == 0 and i != 0:
                print('- - - - - - - - - - -')

            for j in range(9):
                if j % 3 == 0 and j != 0:
                    print('| ', end='')

                if j == 8:
                    print(self.board[i][j])
                else:
                    print(f'{self.board[i][j]} ', end='')
        print('')


if __name__ == "__main__":
    game = Sudoku(game)
    game.print_board()
    game.solve()
    game.print_board()
